### PR TITLE
Clarified documentation of `isless`

### LINKED ISSUE #31311: Usage of `isless`
Dear all,

I think there is a mismatch between the documentation and implementation of `isless`. I am not sure which one should be correct. The documentation states
```
  isless(x, y)

  Test whether x is less than y, according to a canonical total order. Values that are normally unordered, such as NaN, are ordered in an arbitrary but consistent fashion. missing values are ordered last.

  This is the default comparison used by sort.
```

This suggests that any two objects are always comparable, and the result is consistent. However, most comparisons fail with a method error, e.g.,
```
julia> isless(1,[2,3])
ERROR: MethodError: no method matching isless(::Int64, ::Array{Int64,1})    
```
What is the intended behaviour of `isless`. If the intended behaviour of `isless` is to only compare objects of the same "typeclass", I would say the documentation is misleading. For reference, my current Julia version is 1.1.0 release:
```
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.1.0 (2019-01-21)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/  
```

### PR DESCRIPTION
**Pull request by sebasguts**
This PR is a follow up to the discussion in #31311 
Resolves #31311

### CODE REVIEW

File: base/operators.jl [hunk 1]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
 """
     isless(x, y)
 
-Test whether `x` is less than `y`, according to a canonical total order. Values that are
-normally unordered, such as `NaN`, are ordered in an arbitrary but consistent fashion.
+Test whether `x` is less than `y`, according to a fixed total order.
+`isless` is not defined on any pair of values `(x, y)`. However, if it
+is defined, it satisfies the following:
+- If `isless(x, y) is defined, then so is `isless(y, x)` and `x == y`,
```
Comment (chethega) [inline base/operators.jl]: >If `isless(x, y) is defined, then so is `isless(y, x)` and `x == y`,

If `isless(x, y) is defined, then so is `isless(y, x)` and `isequal(x,y)`,

Comment (sebasguts) [reply base/operators.jl]: Thx, I fixed that.

**Comment by chethega** [comment]
I'd like to note that "if `isless(x,y)===true` and `isless(y,z)===true` then `isless(x,z)` is well-defined and true" is probably correct but not entirely trivial: First, tuples containing incomparable elements (`(1,"a"), (1, "b"), (2, 1)`) and second recursive datastructures where `isless`, `isequal`, etc may fail to terminate (there, the claim is: if `isless(x,y)`, `isless(x,z)` terminate and return true, then `isless(x,z)` will not run into an endless loop).

File: base/operators.jl [hunk 2]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
 """
     isless(x, y)
 
-Test whether `x` is less than `y`, according to a canonical total order. Values that are
-normally unordered, such as `NaN`, are ordered in an arbitrary but consistent fashion.
+Test whether `x` is less than `y`, according to a fixed total order.
+`isless` is not defined on any pair of values `(x, y)`. However, if it
+is defined, it satisfies the following:
```
Comment (fingolfin) [inline base/operators.jl]:
```suggestion
is defined, it is expected to satisfy the following:
```

File: base/operators.jl [hunk 3]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
...
+- If `isless(x, y) is defined, then so is `isless(y, x)` and `isequal(x, y)`,
```
Comment (fingolfin) [inline base/operators.jl]:
```suggestion
- If `isless(x, y)` is defined, then so is `isless(y, x)` and `isequal(x, y)`,
```

File: base/operators.jl [hunk 4]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
...
+  and only one of those three yields `true`.
```
Comment (fingolfin) [inline base/operators.jl]:
```suggestion
  and exactly one of those three yields `true`.
```

File: base/operators.jl [hunk 5]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
...
+- The relation defined by `isless` is transitive, i.e.,
+  `isless(x, y) && isless(y, z)` => `isless(x, z)`.
```
Comment (fingolfin) [inline base/operators.jl]:
```suggestion
  `isless(x, y) && isless(y, z)` implies `isless(x, z)`.
```

**Review by fingolfin — COMMENTED** [review]
To me this is a clear improvement, I only have a few nitpicks

File: base/operators.jl [hunk 6]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
...
+are ordered in an arbitrary but consistent fashion.
 [`missing`](@ref) values are ordered last.
 
 This is the default comparison used by [`sort`](@ref).
 
 # Implementation
-Non-numeric types with a canonical total order should implement this function.
+Non-numeric types with a total order should implement this function.
 Numeric types only need to implement it if they have special values such as `NaN`.
 Types with a canonical partial order should implement [`<`](@ref).
```
Comment (JeffBezanson) [inline base/operators.jl]:
```suggestion
Types with a partial order should implement [`<`](@ref).
```

File: base/operators.jl [hunk 7]
```diff
@@ -132,14 +132,22 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
 """
     isless(x, y)
 
-Test whether `x` is less than `y`, according to a canonical total order. Values that are
-normally unordered, such as `NaN`, are ordered in an arbitrary but consistent fashion.
+Test whether `x` is less than `y`, according to a fixed total order.
+`isless` is not defined on any pair of values `(x, y)`. However, if it
```
Comment (JeffBezanson) [inline base/operators.jl]:
```suggestion
`isless` is not defined on all pairs of values `(x, y)`. However, if it
```

**Comment by sebasguts** [comment]
Thank you very much for the clarifications.

@chethega You are right, sometimes such comparisons might be hard to implement. @fingolfin has taken that into account in his changes, I think.

File: base/operators.jl [hunk 8]
```diff
@@ -132,16 +132,24 @@ isequal(x::AbstractFloat, y::Real         ) = (isnan(x) & isnan(y)) | signequal(
...
 [`missing`](@ref) values are ordered last.
 
 This is the default comparison used by [`sort`](@ref).
 
 # Implementation
-Non-numeric types with a canonical total order should implement this function.
+Non-numeric types with a total order should implement this function.
 Numeric types only need to implement it if they have special values such as `NaN`.
-Types with a canonical partial order should implement [`<`](@ref).
+ Types with a partial order should implement [`<`](@ref).
```
Comment (JeffBezanson) [inline base/operators.jl]:
```suggestion
Types with a partial order should implement [`<`](@ref).
```
Sorry, I don't know what happened with this space character.

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**